lib/uicc_lwm2m/pkcs15_decode.c

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+
+#include "asn1_decode.h"
+#include "pkcs15_decode.h"
+
+static void asn1_dec_Path(asn1_ctx_t *ctx, void *data)
+{
+	pkcs15_object_t *object = (pkcs15_object_t *)data;
+	uint8_t tag;
+	size_t len;
+
+	/* Here only path is handled.
+	 *
+	 * PKCS15Path ::= SEQUENCE {
+	 *     path    OCTET STRING,
+	 *     index   INTEGER (0..pkcs15-ub-index) OPTIONAL,
+	 *     length  [0] INTEGER (0..pkcs15-ub-index) OPTIONAL
+	 * } (WITH COMPONENTS {..., index PRESENT, length PRESENT} |
+	 *    WITH COMPONENTS {..., index ABSENT, length ABSENT})
+	 */
+
+	while (asn1_dec_head(ctx, &tag, &len)) {
+		switch (tag) {
+		case UP4:
+			asn1_dec_octet_string(ctx, len, object->path, sizeof(object->path));
+			break;
+		default:
+			asn1_dec_skip(ctx, len);
+			break;
+		}
+	}
+}
+
+static void asn1_dec_PathOrObjects_Data(asn1_ctx_t *ctx, void *data)
+{
+	pkcs15_object_t *object = (pkcs15_object_t *)data;
+	uint8_t tag;
+	size_t len;
+
+	/* Here only the PKCS15Path choice is handled.
+	 *
+	 * PKCS15PathOrObjects {ObjectType} ::= CHOICE {
+	 *     path     PKCS15Path,
+	 *     objects  [0] SEQUENCE OF ObjectType,
+	 *     ...,
+	 *     indirect-protected [1] ReferencedValue {EnvelopedData {SEQUENCE OF ObjectType}},
+	 *     direct-protected [2] EnvelopedData {SEQUENCE OF ObjectType},
+	 * }
+	 */
+
+	while (asn1_dec_head(ctx, &tag, &len)) {
+		switch (tag) {
+		case UC16: /* PKCS15Path */
+			asn1_dec_sequence(ctx, len, &object->path, asn1_dec_Path);
+			break;
+		default:
+			asn1_dec_skip(ctx, len);
+			break;
+		}
+	}
+}
+
+static void asn1_dec_OidDO(asn1_ctx_t *ctx, void *data)
+{
+	pkcs15_object_t *object = (pkcs15_object_t *)data;
+	uint8_t tag;
+	size_t len;
+
+	/* Here only decode value when OID match 2.23.43.9.1 "lwm2m_bootstrap".
+	 *
+	 * Because of a bug in the OMA-TS-LwM2M_Core spec example this
+	 * variant must also be checked.
+	 *
+	 * PKCS15OidDO ::= SEQUENCE {
+	 *     id     OBJECT IDENTIFIER,
+	 *     value  ObjectValue {PKCS15-OPAQUE.&Type}
+	 * }
+	 */
+
+	uint8_t oid[32];
+	bool decode_path = false;
+
+	while (asn1_dec_head(ctx, &tag, &len)) {
+		switch (tag) {
+		case UP6: /* id */
+			/* Using octet string decoder to simplify code using string matching */
+			asn1_dec_octet_string(ctx, len, oid, sizeof(oid));
+
+			/* Check for 2.23.43.9.1 or the OMA-TS variant */
+			if ((strcmp(oid, "672b0901") == 0) || (strcmp(oid, "0604672b0901") == 0)) {
```

Comment:
defines?

Reply:
Added `ASN1_LWM2M_BOOTSTRAP_OID` and `ASN1_OMA_TS_LWM2M_BOOTSTRAP_OID` for the two OID variants.